### PR TITLE
fix: tighten `deepEqual`/`shallowEqual` typings, fix corner-case false positives

### PR DESCRIPTION
**Pull request by agurtovoy**
Our `deepEqual`/`shallowEqual` utils were typed too broadly for their intended plain-data semantics, had no tests, and contained a cross-shape bug where arrays and objects with compatible enumerable keys could compare as equal.

See inline comments for details. Best reviewed commit-by-commit.

### CODE REVIEW

File: packages/controls/src/lib/predicates/__tests__/shallow-deep-equal.test.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
+import { deepEqual } from '../deepEqual'
+import { shallowEqual } from '../shallowEqual'
```
Comment (agurtovoy) [inline packages/controls/src/lib/predicates/__tests__/shallow-deep-equal.test.ts:2]: Since the test cases for `deepEqual` & `shallowEqual` are mostly the same, I'm testing both of them at the same time.

File: packages/controls/src/lib/predicates/__tests__/shallow-deep-equal.test.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      [null, 0],
+      [false, 0],
+      [false, true],
+      [0, -0],
+      [NaN, Infinity],
+      [Infinity, -Infinity],
+      [[], {}],
+      [[17], { '0': 17 }],
+      [{ a: [17] }, { a: { '0': 17 } }],
+      [[null, undefined], { '0': null, '1': undefined }],
```
Comment (agurtovoy) [inline packages/controls/src/lib/predicates/__tests__/shallow-deep-equal.test.ts:53]: The three test cases above (starting from `[[17], { '0': 17 }]`) previously failed as the comparison erroneously returned true.

File: packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx [hunk 3]
```diff
@@ -19,13 +19,18 @@ export function useSyncWithBuilder(editor: Editor, text?: RichTextDTO) {
     if (shouldCommit && text && isInBuilder) {
       const nextValue = richTextDTOtoDAO(text)
       const nextSelection = richTextDTOtoSelection(text)
-      if (!deepEqual(editor.children, nextValue) || !deepEqual(editor.selection, nextSelection)) {
```
Comment (agurtovoy) [inline packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx:22]: `deepEqual` call on Slate's selection instance, which is a `Range` with its own domain-specific comparison semantics.

File: packages/controls/src/lib/predicates/deepEqual.ts [hunk 4]
```diff
@@ -1,8 +1,13 @@
...
+
 import { shallowEqual } from './shallowEqual'
 
 const { hasOwnProperty } = Object.prototype
 
-export const deepEqual = (a: unknown, b: unknown): boolean => {
+/**
+ * Deep equality for scalars, arrays and plain objects.
+ */
+export const deepEqual = (a: Data, b: Data): boolean => {
```
Comment (agurtovoy) [inline packages/controls/src/lib/predicates/deepEqual.ts:10]: Both `deepEqual` and `shallowEqual` were written for and have well-defined, unsurprising behavior only when invoked on plain (recursive) data structures. 

For example, `deepEqual` returns `true` for something as simple as 

```ts
deepEqual(
    { date: new Date("01/01/1999") }, 
    { date: new Date("12/31/1999") }) // true!
````

The `unknown` typings obscured this boundary and encouraged calls on arbitrary objects (as evidenced by [this usage](https://github.com/makeswift/makeswift/pull/1362/changes/6ddaf4be993f98b05b6b20a0f11895062ee2c66d#r3591767868)).

Re-type both as accepting `Data`, our baseline plain-data contract across the `runtime` codebase.

File: packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx [hunk 5]
```diff
@@ -19,13 +19,18 @@ export function useSyncWithBuilder(editor: Editor, text?: RichTextDTO) {
...
+          : Range.equals(editor.selection, nextSelection)
+
+      if (!deepEqual(editor.children, nextValue) || !sameSelection) {
         editor.children = nextValue
         editor.selection = nextSelection
         editor.onChange()
       }
     }
-  }, [editor, shouldCommit, text])
+  }, [editor, isInBuilder, shouldCommit, text])
```
Comment (agurtovoy) [inline packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx:33]: Add a missing hook dependency.

File: packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx [hunk 6]
```diff
@@ -19,13 +19,18 @@ export function useSyncWithBuilder(editor: Editor, text?: RichTextDTO) {
     if (shouldCommit && text && isInBuilder) {
       const nextValue = richTextDTOtoDAO(text)
       const nextSelection = richTextDTOtoSelection(text)
-      if (!deepEqual(editor.children, nextValue) || !deepEqual(editor.selection, nextSelection)) {
+      const sameSelection =
+        editor.selection == null || nextSelection == null
+          ? editor.selection == nextSelection
+          : Range.equals(editor.selection, nextSelection)
```
Comment (agurtovoy) [inline packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx:25]: Use `Range.equals` to compare non-null selections (this is in RichText v1 and thus is mostly inconsequential).

File: packages/runtime/src/slate/TypographyPlugin/normalizeTypographyUp.ts [hunk 7]
```diff
@@ -1,12 +1,12 @@
...
-import { RichTextTypography } from '@makeswift/controls'
+import { type Data, RichTextTypography } from '@makeswift/controls'
 
 import { MakeswiftEditor } from '../types'
 import keys from '../../utils/keys'
 import deepEqual from '../../utils/deepEqual'
 import { shallowMergeTypographies } from './normalizeTypographyDown'
 
-function shallowAnd<A extends Record<string, unknown>, B extends Record<string, unknown>>(
+function shallowAnd<A extends Record<string, Data>, B extends Record<string, Data>>(
```
Comment (agurtovoy) [inline packages/runtime/src/slate/TypographyPlugin/normalizeTypographyUp.ts:9]: Here and below, downstream typing changes.

File: packages/controls/src/lib/predicates/deepEqual.ts [hunk 8]
```diff
@@ -18,6 +18,8 @@ export const deepEqual = (a: Data, b: Data): boolean => {
   )
     return false
 
+  if (Array.isArray(a) !== Array.isArray(b)) return false
```
Comment (agurtovoy) [inline packages/controls/src/lib/predicates/deepEqual.ts:21]: Fix failing tests: https://github.com/makeswift/makeswift/pull/1362/changes/f4bf8862ad5436776059ca9c5bc42c0899b8384a#r3591706380

File: packages/controls/src/lib/predicates/is.ts [hunk 9]
```diff
@@ -1,5 +0,0 @@
-export function is(x: unknown, y: unknown): boolean {
-  if (x === y) return x !== 0 || y !== 0 || 1 / x === 1 / y
-
-  return x !== x && y !== y
-}
```
Comment (agurtovoy) [inline packages/controls/src/lib/predicates/is.ts:5]: Replace the hand-rolled `is` helper with [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is), which has equivalent semantics and has been broadly supported across browsers for over a decade.

File: packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx (hunk 6 above)
Comment (arvinpoddar) [inline packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx:25]: How did you stumble upon this?

**Review by arvinpoddar — APPROVED** [review]
🙏

**Review by bc-drew-cutchins — APPROVED** [review]
Looks good! Thanks for the thorough tests!

File: packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx (hunk 6 above)
Comment (agurtovoy) [reply packages/runtime/src/runtimes/react/controls/rich-text/EditableText/useSyncWithBuilder.tsx:25]: @arvinpoddar It stopped compiling after I tightened up the typings because `Range` is an actual interface, not just a bag of props.
